File: Observer/Jave/src/Observer/product.java
package Observer;

import java.util.ArrayList;
import java.util.List;
import java.util.Observer;

public class product implements  subject{
    private String name;
    private String availability;
    private List<observer> observersList;
    public product(String name){
        this.name = name;
        this.observersList = new ArrayList<>();
    }

    @Override
    public void add(observer observ) {
        this.observersList.add(observ);
    }

    @Override
    public void remove(observer observ) {
        this.observersList.remove(observ);
    }

    @Override
    public void notifyallObserver() {
        for(observer observ : observersList){
            observ.update(availability);
        }
    }
    public void setAvailability(boolean availbile){
        availability = availbile? "available":"not available";
        notifyallObserver();
    }
}

File: Observer/Python/Product.py
# -*- coding: utf-8 -*-
"""
Created on Sat May  7 16:55:16 2022

@author: MahmoudSaeed
"""

import Subject 

class Product(Subject.Subject):
    def __init__(self,name):
        self.name = name
        self.observerList = []
        self.availability = None
      
    def set_availability(self , available:bool):
        if available == True:
            self.availability = "availabel"
        else:
            self.availability = "not"
         
        self.notifyAllObserver()    
        
     
    def add(self , observer):
        self.observerList.append(observer)
        
    def remove(self , observer):
        if observer not in self.observerList:
            pass
        self.observerList.remove(observer)
        
    def notifyAllObserver(self):
        for i in self.observerList:
            i.update(self.availability)
            
            